### PR TITLE
Partial rollback of #2260

### PR DESCRIPTION
**Pull request by trevorcai**
This has caused training divergences in some LM setups using bfloat16.
Temporarily rolling it back until we better understand this issue.

### CODE REVIEW

**Review by jekbradbury — APPROVED** [review]
I don’t understand what’s going on, but the rollback LGTM. We should look at example inputs where the grads are different, the dtypes used in the optimized HLO, the effects of the allow_excess_precision flag, and/or whether we can reproduce the correct convergence without the complexity of grad(max).

**Comment by trevorcai** [comment]
I did part of the requested analysis. The float32 numerics generally agree for all versions of `log_softmax`. In bfloat16, the `stop_gradient` version is closer to the `f32` numerics than without `stop_gradient` when measured with inputs & cotangents sampled from `jax.random.normal`.

This is surprising, given that the version without `stop_gradient` matches the f32 training curve, while including it causes a divergence. I don't have a strong conclusion for why this may be the case.

I did identify that `grad(log_softmax)` in bf16 is biased to be too high, because `exp` when computed in f32 and casted to bf16 is biased to be too low. The errors in the version without `stop_gradient` oppose this bias; it's possible that this helps during training.

I implemented a `jax.custom_vjp` for `log_softmax` following the TensorFlow gradient definition. It's both faster and closer to the f32 numerics than both the alternatives. This still doesn't match the f32 curves, but is closer than the current version.

Given that we don't understand why the version with worse numerics and slower runtime seems to imitate f32 better in practice, I think we should **not** merge this PR. Everyone using f32 benefits from current HEAD, and the benefits for bf16 folks are unclear.

I've worked around this by just computing `log_softmax` in f32.

The aforementioned `custom_vjp` gives a small speedup and reduction in activations stored. If it can be rewritten as a `custom_jvp` (not sure how), I'd recommend adding it to HEAD:

```
@jax.custom_vjp
def log_softmax(x):
  return jax.nn.log_softmax(x)

def lsm_fwd(x):
  out = log_softmax(x)
  return out, out

def lsm_bwd(res, g):
  out = res
  dx = g - jnp.sum(g, axis=-1, keepdims=True) * jnp.exp(out)
  return (dx,)

log_softmax.defvjp(lsm_fwd, lsm_bwd)
```
